Libraries/LibTLS/Exchange.cpp

```diff
@@ -0,0 +1,216 @@
+
+    m_context.crypto.created = 1;
+
+    return true;
+}
+
+void TLSv12::pseudorandom_function(ByteBuffer& output, const ByteBuffer& secret, const u8* label, size_t label_length, const ByteBuffer& seed, const ByteBuffer& seed_b)
+{
+    if (!secret.size()) {
+        dbg() << "null secret";
+        return;
+    }
+
+    // RFC 5246: "In this section, we define one PRF, based on HMAC.  This PRF with the
+    //            SHA-256 hash function is used for all cipher suites defined in this
+    //            document and in TLS documents published prior to this document when
+    //            TLS 1.2 is negotiated."
+    // Apparently this PRF _always_ uses SHA256
+    Crypto::Authentication::HMAC<Crypto::Hash::SHA256> hmac(secret);
+
+    auto l_seed_size = label_length + seed.size() + seed_b.size();
+    u8 l_seed[l_seed_size];
+    auto label_seed_buffer = ByteBuffer::wrap(l_seed, l_seed_size);
+    label_seed_buffer.overwrite(0, label, label_length);
+    label_seed_buffer.overwrite(label_length, seed.data(), seed.size());
+    label_seed_buffer.overwrite(label_length + seed.size(), seed_b.data(), seed_b.size());
+
+    auto digest_size = hmac.digest_size();
+
+    u8 digest[digest_size];
+
+    auto digest_0 = ByteBuffer::wrap(digest, digest_size);
+
+    digest_0.overwrite(0, hmac.process(label_seed_buffer).immutable_data(), digest_size);
+
+    size_t index = 0;
+    while (index < output.size()) {
+        hmac.update(digest_0);
+        hmac.update(label_seed_buffer);
+        auto digest_1 = hmac.digest();
+
+        auto copy_size = min(digest_size, output.size() - index);
+
+        output.overwrite(index, digest_1.immutable_data(), copy_size);
+        index += copy_size;
+
+        digest_0.overwrite(0, hmac.process(digest_0).immutable_data(), digest_size);
+    }
+}
+
+bool TLSv12::compute_master_secret(size_t length)
+{
+    if (m_context.premaster_key.size() == 0 || length < 48) {
+        dbg() << "there's no way I can make a master secret like this";
+        dbg() << "I'd like to talk to your manager about this length of " << length;
+        return false;
+    }
+
+    m_context.master_key.clear();
+    m_context.master_key.grow(length);
+
+    pseudorandom_function(
+        m_context.master_key,
+        m_context.premaster_key,
+        (const u8*)"master secret", 13,
+        ByteBuffer::wrap(m_context.local_random, 32),
+        ByteBuffer::wrap(m_context.remote_random, 32));
+
+    m_context.premaster_key.clear();
+#ifdef TLS_DEBUG
+    dbg() << "master key:";
+    print_buffer(m_context.master_key);
+#endif
+    expand_key();
+    return true;
+}
+
+ByteBuffer TLSv12::build_certificate()
+{
+    dbg() << "FIXME: build_certificate";
+    return {};
+}
+
+ByteBuffer TLSv12::build_change_cipher_spec()
+{
+    PacketBuilder builder { MessageType::ChangeCipher, m_context.version, 64 };
+    builder.append((u8)1);
```

Comment:
That is a magic value, but the ChangeCipher spec only allows a value of 1 so I guess it's OK to not create a constant.

Reply:
I'll just leave it as-is then.